src/update.rs

```diff
@@ -0,0 +1,167 @@
+
+/// An update guard over the [`Enr`].
+/// The inverses are set as a generic to allow optimizing for single updates, multiple updates with
+/// a known count of updates and arbitrary updates.
+pub struct Guard<'a, K: EnrKey, Up: UpdatesT> {
+    /// Testing keep a clone of the enr to verify it remains unchanged on failure.
+    #[cfg(test)]
+    enr_backup: Enr<K>,
+    /// [`Enr`] with update operations already applied.
+    enr: &'a mut Enr<K>,
+    /// Inverses that would need to be applied to the [`Enr`] to restore [`Enr::content`].
+    ///
+    /// Inverses must be in the order in which they were obtained, so that applying them in
+    /// reserved order produces the original content.
+    inverses: Up::ValidatedUpdates,
+}
+
+impl<'a, K: EnrKey, Up: UpdatesT> Guard<'a, K, Up> {
+    /// Create a new guard verifying the update and applying it to the the [`Enr`].
+    ///
+    /// If validation fails, an error is returned and it's guaranteed that the [`Enr`] has not been
+    /// changed.
+    pub fn new(enr: &'a mut Enr<K>, updates: Up) -> Result<Self, EnrError> {
+        // validate the update
+        let updates = updates.to_valid()?;
+        #[cfg(test)]
+        let enr_backup = Enr {
+            seq: enr.seq,
+            node_id: enr.node_id,
+            content: enr.content.clone(),
+            signature: enr.signature.clone(),
+            phantom: std::marker::PhantomData,
+        };
+        // apply the valid operation to the enr and create the inverse
+        let inverses = updates.apply_and_invert(enr);
+        Ok(Self {
+            #[cfg(test)]
+            enr_backup,
+            enr,
+            inverses,
+        })
+    }
+
+    /// Applies the remaining operations in a valid [`Enr`] update:
+    ///
+    /// 1. Add the public key matching the signing key to the contents.
+    /// 2. Update the sequence number.
+    /// 3. Sign the [`Enr`].
+    /// 4. Verify that the encoded [`Enr`] is within spec lengths.
+    /// 5. Update the cache'd node id
+    ///
+    /// If any of these steps fails, all successful changes are reverted.
+    pub fn finish(
+        self,
+        signing_key: &K,
+    ) -> Result<<Up::ValidatedUpdates as ValidUpdatesT>::Output, EnrError> {
+        let Guard {
+            #[cfg(test)]
+            enr_backup,
+            enr,
+            inverses,
+        } = self;
+        let mut revert = RevertOps::new(inverses);
+
+        // 1. set the public key
+        let public_key = signing_key.public();
+        let encoded_pk = rlp::encode(&public_key.encode().as_ref()).freeze();
+        let pk_name = public_key.enr_key();
+        let prev_pk_contents = enr.content.insert(public_key.enr_key(), encoded_pk);
+        revert.public_key = Some((pk_name, prev_pk_contents));
+
+        // 2. set the new sequence number
+        let Some(new_seq) = enr.seq.checked_add(1) else {
+            revert.recover(enr);
+            #[cfg(test)]
+            assert_eq!(&enr_backup, enr);
```

Comment:
We should be careful with these. The Eq implementation doesn't check the content. Just the seq, node-id and signature. We're assuming the sig covers the contents, but in this case, we are just setting it to some value. 

I think its fine, just pointing it out